Rename vmdk driver as vsphere.

govint opened the pull request:
Updated docs and the plugin init code to expose a "vsphere" driver.

plugin logs

2017-02-16 01:42:49.690079924 -0800 PST [INFO] No config file found. Using defaults.
2017-02-16 01:42:49.690279065 -0800 PST [INFO] Starting plugin log_level=debug config="/etc/docker-volume-vsphere.conf" driver=vsphere
2017-02-16 01:42:49.690325892 -0800 PST [WARNING] Failed to load config file %s: %v/etc/docker-volume-vsphere.confopen /etc/docker-volume-vsphere.conf: no such file or directory
2017-02-16 01:42:49.690346301 -0800 PST [INFO] Plugin options - port=1019
2017-02-16 01:42:49.690389849 -0800 PST [INFO] Getting volume data from unix:///var/run/docker.sock
2017-02-16 01:42:49.763661689 -0800 PST [DEBUG] Docker info: version=1.13.0-rc1, root=/var/lib/docker, OS=Ubuntu 14.04.5 LTS
2017-02-16 01:42:49.764895552 -0800 PST [DEBUG] Found 0 running or paused containers
2017-02-16 01:42:49.765255637 -0800 PST [INFO] Discovered 0 volumes in use.
2017-02-16 01:42:49.765306372 -0800 PST [INFO] Docker Vsphere plugin started version="VSphere Volume Driver v0.3" port=1019 mock_esx=false
2017-02-16 01:42:49.765423389 -0800 PST [INFO] Going into ServeUnix - Listening on Unix socket address="/run/docker/plugins/vsphere.sock"
2017-02-16 01:42:49.765706329 -0800 PST [DEBUG] root group found. gid: 0

Docker reports the new driver.

$ docker volume ls
DRIVER              VOLUME NAME
vsphere             newvol1@bigone
vsphere             newvol2@bigone
vsphere             newvol3@bigone
vsphere             newvol@bigone

pdhamdhere commented on docs/misc/plugin-lifecycle.md:
```diff
@@ -22,7 +22,7 @@ A plugin's lifecycle includes.
 
 1. Install: If it is the first install Docker can be running while the plugin is installed. During installs that are part of an upgrade, the admin must insure that docker is stopped and is started after the plugin starts up.
-2. Uninstall: During uninstall docker should be stopped. Before uninstall, make sure no containers still use the VMDK volumes, stop docker engine, and then uninstall the plugin. Docker will generate timeouts on docker start if there are still VMDK volumes after the plugin is uninstalled.
+2. Uninstall: During uninstall docker should be stopped. Before uninstall, make sure no containers still use the Vsphere volumes, stop docker engine, and then uninstall the plugin. Docker will generate timeouts on docker start if there are still Vsphere volumes after the plugin is uninstalled.
```
Vsphere => "vSphere" that seems to standard vmware convention. Please find & replace at other places in change.

pdhamdhere commented on vmdk_plugin/drivers/vmdk/vmdk_driver.go:
```diff
@@ -41,8 +41,8 @@ const (
-	version          = "VMDK Volume Driver v0.3"
-	driverName       = "vmdk"
+	version          = "VSphere Volume Driver v0.3"
+	driverName       = "vsphere"
```
I think we should keep, "vmdk" for backward compatibility. If possible warn user that "vmdk" driver type is deprecated and use "vSphere" instead.

pdhamdhere commented on vmdk_plugin/main.go:
```diff
@@ -38,7 +38,7 @@ const (
 	mountRoot     = "/mnt/vmdk" // VMDK and photon volumes are mounted here
 	photonDriver  = "photon"
-	vmdkDriver    = "vmdk"
+	vmdkDriver    = "vsphere"
```
Looks like driverName is case sensitive. 
```
docker volume create -d VMDK --name=v1@bad
Error response from daemon: create v1@bad: create v1@bad: Error looking up volume plugin VMDK: plugin not found
```

"vSphere" with uppercase S is error prone. Can we register multiple times - "vsphere", "vSphere", "VSPHERE". Or a PR to Docker to make plugin name case insensitive?

govint commented:
Can we instead use "vsphere", because none of the docker command options
are upper case and an incorrect name will return error. Plus Photon driver
is also specified as "photon", or we'll have to change that to Photon.

On Fri, Feb 17, 2017 at 8:05 AM, Prashant Dhamdhere <
notifications@github.com> wrote:

> *@pdhamdhere* commented on this pull request.
> ------------------------------
>
> In docs/misc/plugin-lifecycle.md
> <https://github.com/vmware/docker-volume-vsphere/pull/919#discussion_r101675089>
> :
>
> > @@ -22,7 +22,7 @@ A plugin's lifecycle includes.
>  Note: Upgrade of the plugin is done as an uninstall of the older version and an install of the newer version.
>
>  1. Install: If it is the first install Docker can be running while the plugin is installed. During installs that are part of an upgrade, the admin must insure that docker is stopped and is started after the plugin starts up.
> -2. Uninstall: During uninstall docker should be stopped. Before uninstall, make sure no containers still use the VMDK volumes, stop docker engine, and then uninstall the plugin. Docker will generate timeouts on docker start if there are still VMDK volumes after the plugin is uninstalled.
> +2. Uninstall: During uninstall docker should be stopped. Before uninstall, make sure no containers still use the Vsphere volumes, stop docker engine, and then uninstall the plugin. Docker will generate timeouts on docker start if there are still Vsphere volumes after the plugin is uninstalled.
>
> Vsphere => "vSphere" that seems to standard vmware convention. Please find
> & replace at other places in change.
> ------------------------------
>
> In vmdk_plugin/drivers/vmdk/vmdk_driver.go
> <https://github.com/vmware/docker-volume-vsphere/pull/919#discussion_r101675301>
> :
>
> > @@ -41,8 +41,8 @@ const (
>  	devWaitTimeout   = 1 * time.Second
>  	sleepBeforeMount = 1 * time.Second
>  	watchPath        = "/dev/disk/by-path"
> -	version          = "VMDK Volume Driver v0.3"
> -	driverName       = "vmdk"
> +	version          = "VSphere Volume Driver v0.3"
> +	driverName       = "vsphere"
>
> I think we should keep, "vmdk" for backward compatibility. If possible
> warn user that "vmdk" driver type is deprecated and use "vSphere" instead.
> ------------------------------
>
> In vmdk_plugin/main.go
> <https://github.com/vmware/docker-volume-vsphere/pull/919#discussion_r101675554>
> :
>
> > @@ -38,7 +38,7 @@ const (
>  	pluginSockDir = "/run/docker/plugins"
>  	mountRoot     = "/mnt/vmdk" // VMDK and photon volumes are mounted here
>  	photonDriver  = "photon"
> -	vmdkDriver    = "vmdk"
> +	vmdkDriver    = "vsphere"
>
> Looks like driverName is case sensitive.
>
> docker volume create -d VMDK --name=v1@bad
> Error response from daemon: create v1@bad: create v1@bad: Error looking up volume plugin VMDK: plugin not found
>
> "vSphere" with uppercase S is error prone. Can we register multiple times
> - "vsphere", "vSphere", "VSPHERE". Or a PR to Docker to make plugin name
> case insensitive?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/919#pullrequestreview-22411685>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APHseNg_VLmq1G3zZWZPuOJcwzTaPZZPks5rdQdlgaJpZM4MC1Z5>
> .
>

pdhamdhere commented:
>Can we instead use "vsphere", because none of the docker command options

Sure.

shaominchen commented:
I agree with Mark and Prashant that we should keep backward compatibility for the "vmdk" driver name, and warn user it is deprecated. As for the case sensitivity of the plugin name, I would suggest to favor local consistency - since docker is using lower case consistently, we should just follow that.

govint commented:
Thanks for reviewing. Updated diffs based on comments.

1. Corrected to use vSphere when logging driver init.
2. Updated docs to indicate the two names for the vsphere volume plugins and that "vmdk" will be deprecated.
3. Supporting both names for now. But catch is at some point user must stop using the vmdk name.

Release notes for 0.12 release must mention the change name and plan to deprecate by 0.13.

shaominchen commented on docs/user-guide/docker-volume-cli.md:
```diff
@@ -8,33 +8,33 @@ The service works with existing Docker volume commands.
 
 
-The Docker volume commands are supported for both the Vmdk and Photon platforms with minor differences in capabilities. Features that are specific to either of the platforms are mentioned explicitly below.
+The Docker volume commands are supported for both the Vsphere and Photon platforms with minor differences in capabilities. Features that are specific to either of the platforms are mentioned explicitly below.
```
Vsphere => vSphere. For documents/comments, I think we should use "vSphere".

shaominchen commented on docs/user-guide/docker-volume-cli.md:
```diff
@@ -8,33 +8,33 @@ The service works with existing Docker volume commands.
 The default volume size is 100mb
 
-### vsan-policy-name (Vmdk only)
+### vsan-policy-name (Vsphere only)
```
Same as above.

shaominchen commented on docs/user-guide/docker-volume-cli.md:
````diff
@@ -8,33 +8,33 @@ The service works with existing Docker volume commands.
 Policy needs to be created via the vmdkops-admin-cli. Once policy is created, it can be addressed during create by passing the ```-o vsan-policy-name``` flag.
 
-### diskformat (Vmdk only)
+### diskformat (Vsphere only)
````
Same as above.

shaominchen commented on docs/user-guide/docker-volume-cli.md:
```diff
@@ -45,10 +45,10 @@ Currently the following are supported
 3. Thick Provision Eager Zeroed ([eagerzeroedthick]((https://kb.vmware.com/selfservice/microsites/search.do?language=en_US&cmd=displayKC&externalId=1022242)))
 
-### attach_as (Vmdk only)
+### attach_as (Vsphere only)
```
Same as above.

shaominchen commented on docs/user-guide/docker-volume-cli.md:
```diff
@@ -57,33 +57,33 @@ Currently the following are supported
 2. [independent_persistent](http://cormachogan.com/2013/04/16/what-are-dependent-independent-disks-persistent-and-non-persisent-modes/): If the VMDK is attached as independent_persistent it will not be part of a VM snapshot. The Docker volume can be attached to any VM that can access the datastore independent of snapshots.
 
-### access (Vmdk only)
+### access (Vsphere only)
```
Same as above.

shaominchen commented on docs/user-guide/docker-volume-cli.md:
```diff
@@ -57,33 +57,33 @@ Currently the following are supported
 Specifies which filesystem will be created on the new volume. vSphere Docker Volume Service will search for a existing /sbin/mkfs.**fstype** on the docker host to create the filesystem, and if not found it will return a list of filesystems for which it has found a corresponding mkfs. The specified filesystem must be supported by the running kernel and support labels (-L flag for mkfs). Defaults t ...
 
-### clone-from (Vmdk only)
+### clone-from (Vsphere only)
```
Same as above.

shaominchen commented on vmdk_plugin/drivers/vmdk/vmdk_driver.go:
```diff
@@ -15,7 +15,7 @@
 
 //
-// VMWare VMDK Docker Data Volume plugin.
+// VMWare Vsphere Docker Data Volume plugin.
```
Same as above.

shaominchen commented on vmdk_plugin/main.go:
```diff
@@ -151,7 +152,7 @@ func main() {
 		driver = photon.NewVolumeDriver(*targetURL, *projectID,
 			*vmID, mountRoot)
-	} else if *driverName == vmdkDriver {
+	} else if *driverName == vmdkDriver || *driverName == vsphereDriver {
```
Can we print a message to warn user that "vmdk" driver is deprecated (print a warning message, but still continue to run the command)?

govint replied:
Thanks, fixed the comments and the docs and pushed again.

msterin commented on vmdk_plugin/drivers/vmdk/vmdk_driver.go:
```diff
@@ -41,8 +41,8 @@ const (
 	watchPath        = "/dev/disk/by-path"
-	version          = "VMDK Volume Driver v0.3"
-	driverName       = "vmdk"
+	version          = "vSphere Volume Driver v0.3"
```
I think version is overdue for the change

govint replied:
Notes in docs/user-guide/docker-plugin-drivers.md is updated to mention the change and deprecation of the vmdk driver name. 

For version number, which number should we use?

pdhamdhere replied:
v0.4 with this change?

msterin replied:
I'd suggest 0.11 to catch up with the release #. But if you prefer 0.4 to stay sequential it's good too

govint commented:
Updated the change to retain "vmdk" as the default for now (when no args are provided). This is so existing users can continue to use the current name till the "vmdk" name is deprecated. Users can still use "vsphere" if they start the plugin with the "-driver=vsphere" option. Later we will remove the vmdk name all together and make "vsphere" the default.

pdhamdhere commented on vmdk_plugin/main.go:
```diff
@@ -97,8 +98,8 @@ func main() {
 	// Define command line options
-	driverName := flag.String("driver", "vmdk", "Volume driver")
-	logLevel := flag.String("log_level", "debug", "Logging Level")
+	driverName := flag.String("driver", vmdkDriver, "Volume driver")
```
This change expects user to either configure or pass command line argument --driver=vsphere.

Any reason why we can't register both "vmdk" and "vsphere" pointing to same handlers? Ideally, with NO changes in installation process, following commands should work;

`docker volume create -d=vmdk v1`
Not sure if we could postfix warning that vmdk is deprecated and use vsphere instead.
`docker volume create -d=vsphere v1`

This can potentially open volume naming issues which I believe Docker tracks per Volume Driver. However, we don't expect user to mix & match vmdk/vsphere but rather to stick to one.  So, we should be fine.

govint replied:
True, we can run both instances in parallel, only that adds more code that will need to be removed later. Docker on startup associates a driver to each volume that each driver reports. A volume created by a vmdk driver can get reported by a vsphere driver instance later (after docker and plugin restart) and Docker would associate that volume as a vsphere volume.

Only issue is with scripts - test scripts and user scripts to create a docker app stack that specifies the volume driver to use.
I can make driver also a part of the configuration so user can change the driver name there vs. change the startup scripts.

msterin replied:
I also do not think running 2 process is a good idea - but can we just register the driver with the same socket twice ? Or can we register it twice and listen on both sockets from one driver ? Restarting the driver service to switch between supporting `-d vmdk` and `-d vsphere` seems pointless, if that's the only choice then I think we (anbd customers) would be better off if we  make a breaking change and stick with `vsphere`

govint replied:
Its possible to start two listeners on two separate sockets, the sockets are named after the driver - vmdk.sock, vsphere.sock. One instance of the plugin process runs with the two listeners in this case.

govint replied:
I'll try out having one listener run as a goroutine and check that its working ok.

BUt do we need this capability at all?

govint commented:
OK, tried out having both vmdk and vsphere drivers registered (and listening) in a single instance of the plugin. Volume creation is no problem, volume reporting certainly is. The way volume reporting happens is Docker asks each registered plugin what its volumes are and in our case since both vsphere and vmdk instances are registered, both end up reporting the same volumes.

I'm removing the dual driver registration and keeping "vmdk" the default while vsphere can be used if the user wishes.

$ docker volume create --driver=vsphere --name=vsp1
vsp1

$ docker volume create --driver=vmdk --name=vsp2
vsp2

$ docker volume ls
DRIVER              VOLUME NAME
vmdk                vsp1@bigone
vsphere             vsp1@bigone
vmdk                vsp2@bigone
vsphere             vsp2@bigone
vmdk                vsp@bigone
vsphere             vsp@bigone

msterin commented:
> I'm removing the dual driver registration and keeping "vmdk" the default while vsphere can be used if the user wishes.

So 'vmdk' will be by default, and for 'vsphere' admins would need to do something (what is this that they need to do, btw ? )  Is it what we wanted ? I think if we are chaning to vsphere, we should make it default. @pdhamdhere

pdhamdhere commented:
Installation time Config option to choose driver type (vmdk/vsphere) doesn't make sense. If there is no easy way to support both drivers simultaneously (without manual intervention), then let's just switch to "vsphere" and release note in BOLD, CAPS, RED that vmdk is deprecated. 

You will have to combine this with test changes.

govint commented:
Agreed, I'll post the updates.

govint commented:
In which case to reduce impact can we announce this in the 0.12 release as an "up coming change" and actually make it for 0.13.

@pdhamdhere

pdhamdhere commented:
>In which case to reduce impact can we announce this in the 0.12 release as an "up coming change" and actually make it for 0.13.

Not sure how useful would be 30-day advance notice. It's going to be breaking change whenever we do it. We better do it sooner than later.

govint commented:
@pdhamdhere the last updates before this push was to announce that the vmdk
driver deprecated. Are we changing that again

On Feb 24, 2017 11:50 AM, "Prashant Dhamdhere" <notifications@github.com>
wrote:

> *@pdhamdhere* commented on this pull request.
>
> So, post this change, user must use -d vsphere.
>
> If he wants to continue to use -d vmdk until s/he updates scripts then
> startup script should be updated to add --driver=vmdk in command line?
> Can you add an example in PR description and make sure it works.
>
> Is my understanding correct?
>
> We will need to release note this behavior.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/919#pullrequestreview-23655403>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APHseOt5VALeqb0UcCyuA3b4kC_TPNPYks5rfnafgaJpZM4MC1Z5>
> .
>

govint commented:
@pdhamdhere are these changes ok the vmdk driver option is deprecated from
earlier updates and driver version has been update to 0.4

On Feb 24, 2017 12:32 PM, "Govindan" <tgovin1@gmail.com> wrote:

> @pdhamdhere the last updates before this push was to announce that the
> vmdk driver deprecated. Are we changing that again
>
> On Feb 24, 2017 11:50 AM, "Prashant Dhamdhere" <notifications@github.com>
> wrote:
>
>> *@pdhamdhere* commented on this pull request.
>>
>> So, post this change, user must use -d vsphere.
>>
>> If he wants to continue to use -d vmdk until s/he updates scripts then
>> startup script should be updated to add --driver=vmdk in command line?
>> Can you add an example in PR description and make sure it works.
>>
>> Is my understanding correct?
>>
>> We will need to release note this behavior.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/vmware/docker-volume-vsphere/pull/919#pullrequestreview-23655403>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/APHseOt5VALeqb0UcCyuA3b4kC_TPNPYks5rfnafgaJpZM4MC1Z5>
>> .
>>
>

pdhamdhere commented:
Sorry for confusion. How about following?

By default plugin registers & starts with "vsphere" driver type.

If user wants to use "VMDk" then he can configure driver name in config file or pass as startup CLI argument. We need to document this in release note so let's add "how to" in PR description.

govint commented:
That was the code a couple of iterations back. Now will add it back agan.

msterin commented on docs/user-guide/docker-plugin-drivers.md:
```diff
@@ -43,3 +43,5 @@ Note:
 2. The "target port" is set to "9000" if authentication isn't enabled in photon controller, else port "443" is used.
 3. Both project and VM IDs mentioned above can be obtained via the "photon" CLI, see https://github.com/vmware/photon-controller/wiki/Compile-Photon-Controller-CLI.
+4. The plugin driver supports both "vsphere" and "vmdk" as driver names and both can be used interchangeably for now. The "vmdk" driver name will be deprecated in a later release, allowing users to use only "vsphere" as the driver name.
+5. The "vmdk" driver is however the default driver name that is used when the plugin is installed.
```
I did not find the explanation of how to switch between vmdk and vsphere (i.e `edit xxx file; restart the service` or something along these lines

govint commented:
Updated diff please see last diffs. Added Driver as an option for user to specify in the config file. The default in the code is "vsphere", which can be overridden with the "Driver" setting in the config file (if provided or present at default location) and the command line parameter overrides the conf. file setting. This helps us with the Photon platform as well where user can set the Driver option in the config file and the code will handle that. Service start up scripts should not be modified by the user and hence left those as is.

Updated documentation also.

pdhamdhere commented on vmdk_plugin/main.go:
```diff
@@ -97,17 +99,31 @@ func main() {
+
+	// Default driver, overridden by the config file
+	name := vsphereDriver
+	if c.Driver != "" {
```
I am not a go expert. Would dereferencing "Driver" (`c.Driver`) crash if "c" is NULL?

govint replied:
Reordered the code to select the driver if no error in config file, logs below,

2017-02-27 20:54:44.414369896 -0800 PST [INFO] No config file found. Using defaults.
2017-02-27 20:54:44.415016096 -0800 PST [WARNING] Failed to load config file %s: %v/etc/docker-volume-vsphere.confopen /etc/docker-volume-vsphere.conf: no such file or directory
2017-02-27 20:54:44.415259471 -0800 PST [INFO] Starting plugin log_level=info config="/etc/docker-volume-vsphere.conf" driver=vsphere
2017-02-27 20:54:44.415492463 -0800 PST [INFO] Plugin options - port=1019
2017-02-27 20:54:44.415743608 -0800 PST [INFO] Getting volume data from unix:///var/run/docker.sock
2017-02-27 20:54:44.437056322 -0800 PST [INFO] Discovered 0 volumes in use.
2017-02-27 20:54:44.437362054 -0800 PST [INFO] Docker VMDK plugin started mock_esx=false version="vSphere Volume Driver v0.4" port=1019
2017-02-27 20:54:44.437833239 -0800 PST [INFO] Going into ServeUnix - Listening on Unix socket address="/run/docker/plugins/vsphere.sock"

pdhamdhere commented on vmdk_plugin/main.go:
```diff
@@ -97,17 +99,31 @@ func main() {
+	driverName := flag.String("driver", name, "Volume driver")
+
 	// photon driver options
 	targetURL := flag.String("target", "", "Photon controller URL")
```
Not related to this change but if user starts Plugin with "photon" driver then we should fail to start if required parameters are missing in CLI or config file.

govint replied:
Yes, there is a verification step for the photon driver to check the options given and exit.

pdhamdhere commented on vmdk_plugin/main.go:
```diff
@@ -120,13 +136,7 @@ func main() {
-
-	// The vmdk driver doesn't depend on the config file for options.
+	// The vSphere driver doesn't depend on the config file for options
 	if *driverName == photonDriver && err == nil {
```
It's awkward to chck "err == nil" which is set earlier in the function. I think code can be restructured a bit to do validation in beginning.

govint replied:
Done.

govint commented:
Actually, looks much better now than earlier :-)

On Tue, Feb 28, 2017 at 11:21 AM, Prashant Dhamdhere <
notifications@github.com> wrote:

> *@pdhamdhere* approved this pull request.
>
> Thanks for doing this and apologies for back and forth.
>
> LGTM.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/919#pullrequestreview-24164445>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APHseJgnTTXGevd6t280b9Nju4Lzp6G-ks5rg7XHgaJpZM4MC1Z5>
> .
>
